src/wallet/wallet.cpp

```diff
@@ -3032,9 +3024,10 @@ bool CWallet::CreateTransactionInternal(
                             break;
                         }
 
-                        // The change value was not enough to cover the fees, even after dropping the change. Try selecting again.
-                        assert(change_and_fee < nFeeRet);
-                        continue;
+                        // Since we use effective values now, it should not be possible to reach this situation where the change value was not
+                        // enough to cover the fees
+                        error = _("Cannot reduce change to cover transaction fees");
```

Comment:
In commit "Have KnapsackSolver actually use effective values" (a07552b94fe05651970e53813ae67606692d6f19)

Would be good to change the error to "Bug: Cannot reduce change to cover transaction fees" or "Unexpected condition: Cannot reduce change to cover transaction fees" or just use CHECK_NONFATAL(false) or assert(false), so it is clearer this shouldn't happen, or if it does happen it's a bug and not the user doing something wrong.

Reply:
Following your suggested diff, this error condition was removed and I haven't quite figured out the correct place to reintroduce it.

Reply:
In commit "Have KnapsackSolver actually use effective values" (01dc8ebda50a382d45d3d169b2c3f3965869dcae)

> Following your suggested diff, this error condition was removed and I haven't quite figured out the correct place to reintroduce it.

Not important, but I think the updated equivalent after 01dc8ebda50a382d45d3d169b2c3f3965869dcae would be:

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index b229d8ad63c..414c70372b7 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -3004,10 +3004,9 @@ bool CWallet::CreateTransactionInternal(
                     error = _("Signing transaction failed");
                     return false;
                 }
-                nFeeRet = coin_selection_params.m_effective_feerate.GetFee(nBytes);
+                CAmount fee_needed = coin_selection_params.m_effective_feerate.GetFee(nBytes);
 
                 // Subtract fee from the change output if not subtrating it from recipient outputs
-                CAmount fee_needed = nFeeRet;
                 if (nSubtractFeeFromAmount == 0) {
                     change_position->nValue -= fee_needed;
                 }
@@ -3071,6 +3070,9 @@ bool CWallet::CreateTransactionInternal(
                     nFeeRet = fee_needed;
                     break; // The fee has been deducted from the recipients, nothing left to do here
                 }
+                CHECK_NONFATAL(!coin_selection_params.use_bnb);
+                CHECK_NONFATAL(nFeeRet == 0);
+                nFeeRet = fee_needed;
             }
 
             // Give up if change keypool ran out and change is required
```
```